src/main/java/signup_and_login/UserInteractor.java

```diff
@@ -0,0 +1,43 @@
+package signup_and_login;
+import entities.User;
+
+import java.util.List;
+
+public class UserInteractor implements UserInputBoundary {
+    private final UserOutputBoundary presenter;
+    private final UserDBGateway database;
+
+    public UserInteractor(UserOutputBoundary presenter, UserDBGateway database) {
+        this.presenter = presenter;
+        this.database = database;
+    }
+
+    /**
+     * Handles the user request by checking the existing users in the database.
+     * If a matching username and password are found, it prepares the JoinOrHostView through the presenter.
+     * If the username exists but the password doesn't match, it prepares the view indicating invalid credentials.
+     * If the username is not found, it saves a new user to the database and prepares the JoinOrHostView.
+     *
+     * @param request the user model containing the username and password of our requesting user
+     */
+    @Override
+    public void handleUser(UserModel request) {
+        String plainPassword = request.getPassword();
+        request.setPassword(User.hashPassword(plainPassword));
+        List<UserModel> existingUsers = database.loadUsers();
+
+        for (UserModel existingUser : existingUsers) {
+            if (existingUser.getUsername().equals(request.getUsername())) {
```

Comment:
It might be better to check uuid here instead of username.

Reply:
I'm still checking Username so that we don't allow duplicate usernames, effectively this is checking UserID as well because, the relationship of username to UserID is one to one (one user/username per row = userID). If you still want me to explicitly check it I can just let me know why. See commit 447b882605d4d15fc8cb8bd701de977f6cbfa13e for details.